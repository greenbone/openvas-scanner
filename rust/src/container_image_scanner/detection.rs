use std::fmt::Display;

use thiserror::Error;
use tokio::{
    fs::File,
    io::{AsyncBufRead, AsyncBufReadExt, BufReader},
};

use crate::container_image_scanner::{
    ExternalError,
    image::extractor::{Locator, LocatorError},
};

#[derive(Debug)]
pub struct OperatingSystem {
    pub name: String,
    pub version: String,
    pub version_id: String,
}

impl Display for OperatingSystem {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "name: '{}', version_id: '{}'",
            self.name, self.version_id
        )
    }
}

#[derive(Error, Debug)]
enum OperatingSystemDetectionError {
    #[error("Failed to read the operating system file")]
    ReadError(#[from] std::io::Error),
    #[error("Unknown operating system")]
    Unknown,
}

pub const OS_FILES: &[&str] = &["etc/os-release", "usr/lib/os-release"];

pub async fn operating_system<T>(locator: &T) -> Result<OperatingSystem, ExternalError>
where
    T: Locator,
{
    for path in OS_FILES {
        let path = match locator.locate(path).await {
            Ok(x) => x,
            Err(LocatorError::NotFound(_)) => continue,
            Err(x) => return Err(x.into()),
        };
        let osd = OperatingSystemDetector::new(path.open().await);
        let os = osd.detect_operating_system().await?;
        return Ok(os);
    }
    Err(LocatorError::NotFound("No operating system definition".to_owned()).into())
}

struct OperatingSystemDetector<T> {
    reader: T,
}

impl<T> OperatingSystemDetector<T>
where
    T: AsyncBufRead + Unpin,
{
    fn new(reader: T) -> Self {
        OperatingSystemDetector { reader }
    }

    async fn detect_operating_system(
        self,
    ) -> Result<OperatingSystem, OperatingSystemDetectionError> {
        let mut lines = self.reader.lines();
        let mut os_name = None;
        let mut os_version_id = None;
        let mut os_version = None;

        while let Ok(Some(line)) = lines.next_line().await {
            let line = line.trim();
            if line.starts_with("ID=") {
                os_name = Some(
                    line.split('=')
                        .nth(1)
                        .unwrap_or("")
                        .trim_matches('"')
                        .to_string(),
                );
            }
            if line.starts_with("VERSION_ID=") {
                os_version_id = Some(
                    line.split('=')
                        .nth(1)
                        .unwrap_or("")
                        .trim_matches('"')
                        .to_string(),
                );
            }
            if line.starts_with("VERSION=") {
                os_version = Some(
                    line.split('=')
                        .nth(1)
                        .unwrap_or("")
                        .trim_matches('"')
                        .to_string(),
                );
            }
        }
        if os_name.is_none() || os_version_id.is_none() || os_version.is_none() {
            return Err(OperatingSystemDetectionError::Unknown);
        }
        Ok(OperatingSystem {
            name: os_name.unwrap(),
            version_id: os_version_id.unwrap(),
            version: os_version.unwrap(),
        })
    }
}

impl OperatingSystemDetector<BufReader<File>> {
    #[cfg(test)]
    async fn try_open<T>(root: T) -> Result<Self, std::io::Error>
    where
        T: AsRef<std::path::Path>,
    {
        let file = File::open(root).await?;

        let reader = BufReader::new(file);
        Ok(OperatingSystemDetector { reader })
    }
}

impl From<&'static str> for OperatingSystemDetector<&[u8]> {
    fn from(content: &'static str) -> Self {
        let reader = content.as_bytes();
        OperatingSystemDetector { reader }
    }
}

#[cfg(test)]
mod test {

    use tokio::{fs::File, io::AsyncWriteExt};

    use crate::container_image_scanner::detection::OperatingSystemDetector;

    #[tokio::test]
    async fn test_os_detection_ubuntu() {
        let content = r#"PRETTY_NAME="Ubuntu 24.04.2 LTS"
NAME="Ubuntu"
VERSION_ID="24.04"
VERSION="24.04.2 LTS (Noble Numbat)"
VERSION_CODENAME=noble
ID=ubuntu
ID_LIKE=debian"#;
        let detector = OperatingSystemDetector::from(content);
        let os = detector.detect_operating_system().await.unwrap();
        assert_eq!(os.name, "ubuntu");
        assert_eq!(os.version_id, "24.04");
    }

    #[tokio::test]
    async fn test_os_detection_debian() {
        let content = r#"PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
    NAME="Debian GNU/Linux"
    VERSION_ID="12"
    VERSION="12 (bookworm)"
    VERSION_CODENAME=bookworm
    ID=debian"#;
        // let os = OperatingSystem::detect(&fs).await.unwrap();
        let detector = OperatingSystemDetector::from(content);
        let os = detector.detect_operating_system().await.unwrap();
        assert_eq!(os.name, "debian");
        assert_eq!(os.version_id, "12");
    }

    #[tokio::test]
    async fn test_os_detection_open_euler() {
        let content = r#"NAME="openEuler"
    VERSION="24.03 (LTS-SP1)"
    ID="openEuler"
    VERSION_ID="24.03"
    PRETTY_NAME="openEuler 24.03 (LTS-SP1)""#;
        let detector = OperatingSystemDetector::from(content);
        let os = detector.detect_operating_system().await.unwrap();
        assert_eq!(os.name, "openEuler");
        assert_eq!(os.version_id, "24.03");
        assert_eq!(os.version, "24.03 (LTS-SP1)");
    }

    #[tokio::test]
    async fn test_os_detection_from_file() {
        let content = r#"PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
    NAME="Debian GNU/Linux"
    VERSION_ID="12"
    VERSION="12 (bookworm)"
    VERSION_CODENAME=bookworm
    ID=debian"#;
        let mut file = File::create("/tmp/os-release").await.unwrap();
        file.write_all(content.as_bytes()).await.unwrap();
        let detector = OperatingSystemDetector::try_open("/tmp/os-release")
            .await
            .unwrap();
        let os = detector.detect_operating_system().await.unwrap();
        assert_eq!(os.name, "debian");
        assert_eq!(os.version_id, "12");
        assert_eq!(os.version, "12 (bookworm)");
    }
}
