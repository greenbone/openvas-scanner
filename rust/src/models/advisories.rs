// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

use std::collections::HashMap;

/// Represents an advisory json file for notus product.
#[cfg_attr(feature = "serde_support", derive(serde::Deserialize))]
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ProductsAdvisories {
    /// Version of the advisory file
    pub version: String,
    /// SPDX license identifier
    #[cfg_attr(feature = "serde_support", serde(rename = "spdx-license-identifier"))]
    pub license_identifier: String,
    /// Copyright
    pub copyright: String,
    /// Vulnerability Family
    pub family: String,
    /// List of Advisories
    #[cfg_attr(feature = "serde_support", serde(default))]
    pub advisories: Vec<Advisory>,
}

/// Represents an advisory json file for notus product.
#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
pub struct Advisory {
    /// The advisory's title.
    pub title: String,
    /// The advisory's ID.
    pub oid: String,
    /// Creation Date
    pub creation_date: u64,
    /// Last modification date
    pub last_modification: u64,
    /// Advisory ID
    pub advisory_id: String,
    /// Advisory xref
    pub advisory_xref: String,
    /// Advisory contains a CVE that is listed in the catalog of Known Exploited CVEs from CISA
    #[cfg_attr(feature = "serde_support", serde(default))]
    pub cisa_kev: bool,
    /// List of cves
    #[cfg_attr(feature = "serde_support", serde(default))]
    pub cves: Vec<String>,
    /// Summary
    pub summary: String,
    /// Insight
    #[cfg_attr(feature = "serde_support", serde(default))]
    pub insight: String,
    /// Affected
    pub affected: String,
    /// List of xrefs
    #[cfg_attr(feature = "serde_support", serde(default))]
    pub xrefs: Vec<String>,
    /// Quality of detection
    pub qod_type: String,
    /// Severity
    pub severity: Severity,
}

/// A single vulnerability from an advisory file to be stored
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
#[derive(Default, Debug, Clone, PartialEq, Eq)]
pub struct Vulnerability {
    /// VT Parameters
    pub vt_params: Vec<String>,
    /// Creation Date
    pub creation_date: u64,
    /// Last modification date
    pub last_modification: u64,
    /// Summary
    pub summary: String,
    /// Impact
    pub impact: String,
    /// Affected
    pub affected: String,
    /// Insight
    pub insight: String,
    /// Solution
    pub solution: String,
    /// Solution Type
    pub solution_type: String,
    /// Vuldetect
    pub vuldetect: String,
    /// Quality of detection
    pub qod_type: String,
    /// Severity vector
    pub severity_vector: String,
    /// File name
    pub filename: String,
    /// All references: xrefs, cves, xrefs, advisory xrefs and advisory id.
    pub refs: HashMap<String, Vec<String>>,
    /// Vulnerability Family
    pub family: String,
    /// Title
    pub name: String,
    /// Category
    pub category: String,
}

/// Severity
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
pub struct Severity {
    /// Origin of the severity
    pub origin: String,
    /// severity date
    pub date: u64,
    /// Cvss version v2
    #[cfg_attr(
        feature = "serde_support",
        serde(skip_serializing_if = "Option::is_none")
    )]
    pub cvss_v2: Option<String>,
    /// cvss vector v3
    #[cfg_attr(
        feature = "serde_support",
        serde(skip_serializing_if = "Option::is_none")
    )]
    pub cvss_v3: Option<String>,
}

pub struct ProductsAdvisoriesIterator<'a> {
    products_advisories: &'a ProductsAdvisories,
    index: usize,
}

impl<'a> Iterator for ProductsAdvisoriesIterator<'a> {
    type Item = &'a Advisory;

    fn next(&mut self) -> Option<&'a Advisory> {
        if self.index < self.products_advisories.advisories.len() {
            let result = Some(&self.products_advisories.advisories[self.index]);
            self.index += 1;
            result
        } else {
            None
        }
    }
}

impl ProductsAdvisories {
    pub fn iter(&self) -> ProductsAdvisoriesIterator {
        ProductsAdvisoriesIterator {
            products_advisories: self,
            index: 0,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct VulnerabilityData {
    pub adv: Advisory,
    pub family: String,
    pub filename: String,
}

impl From<VulnerabilityData> for Vulnerability {
    fn from(data: VulnerabilityData) -> Self {
        let sv = match data.adv.severity.cvss_v2 {
            Some(cvss) => cvss,
            None => match data.adv.severity.cvss_v3 {
                Some(cvss) => cvss,
                None => "".to_string(),
            },
        };

        let mut refs = HashMap::new();
        let mut url = data.adv.xrefs.clone();
        url.push(data.adv.advisory_xref.clone());
        if data.adv.cisa_kev {
            refs.insert(
                "CISA".to_string(),
                vec!["Known Exploited Vulnerability (KEV) catalog".to_string()],
            );
            url.push("https://www.cisa.gov/known-exploited-vulnerabilities-catalog".to_string());
        }
        refs.insert("URL".to_string(), url);

        Self {
            vt_params: Vec::new(),
            creation_date: data.adv.creation_date,
            last_modification: data.adv.last_modification,
            summary: data.adv.summary,
            impact: "".to_string(),
            affected: data.adv.affected,
            insight: data.adv.insight,
            solution: "Please install the updated package(s).".to_string(),
            solution_type: "VendorFix".to_string(),
            vuldetect: "Checks if a vulnerable package version is present on the target host."
                .to_string(),
            qod_type: data.adv.qod_type,
            severity_vector: sv,
            filename: data.filename,
            refs,
            family: data.family,
            name: data.adv.title,
            category: "3".to_string(),
        }
    }
}
