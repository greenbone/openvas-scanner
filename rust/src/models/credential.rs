// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

/// Represents a set of credentials to be used for scanning to access a host.
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
pub struct Credential {
    /// Service to use for accessing a host
    pub service: Service,
    /// Port used for getting access. If missing a standard port is used
    pub port: Option<u16>,
    #[cfg_attr(feature = "serde_support", serde(flatten))]
    /// Type of the credential to get access. Different services support different types.
    pub credential_type: CredentialType,
}

impl Credential {
    /// Maps the password of the credential using the given closure.
    pub fn map_password<F, E>(self, f: F) -> Result<Self, E>
    where
        F: Fn(String) -> Result<String, E>,
    {
        Ok(Credential {
            service: self.service,
            port: self.port,
            credential_type: self.credential_type.map_password(f)?,
        })
    }
}

impl Default for Credential {
    fn default() -> Self {
        Self {
            service: Service::SSH,
            port: Default::default(),
            credential_type: CredentialType::UP {
                username: "root".to_string(),
                password: "".to_string(),
                privilege: None,
            },
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
pub struct PrivilegeInformation {
    #[cfg_attr(feature = "serde_support", serde(rename = "privilege_username"))]
    pub username: String,
    #[cfg_attr(feature = "serde_support", serde(rename = "privilege_password"))]
    pub password: String,
}

/// Enum of available services
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
pub enum Service {
    #[cfg_attr(feature = "serde_support", serde(rename = "ssh"))]
    /// SSH, supports [UP](CredentialType::UP) and [USK](CredentialType::USK) as credential types
    SSH,
    #[cfg_attr(feature = "serde_support", serde(rename = "smb"))]
    /// SMB, supports [UP](CredentialType::UP)
    SMB,
    #[cfg_attr(feature = "serde_support", serde(rename = "esxi"))]
    /// ESXi, supports [UP](CredentialType::UP)
    ESXi,
    #[cfg_attr(feature = "serde_support", serde(rename = "snmp"))]
    /// SNMP, supports [SNMP](CredentialType::SNMP)
    SNMP,
    #[cfg_attr(feature = "serde_support", serde(rename = "krb5"))]
    /// SNMP, supports [SNMP](CredentialType::SNMP)
    KRB5,
}

impl AsRef<str> for Service {
    fn as_ref(&self) -> &str {
        match self {
            Service::SSH => "ssh",
            Service::SMB => "smb",
            Service::ESXi => "esxi",
            Service::SNMP => "snmp",
            Service::KRB5 => "krb5",
        }
    }
}

impl TryFrom<&str> for Service {
    type Error = String;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        Ok(match value {
            "ssh" => Service::SSH,
            "smb" => Service::SMB,
            "esxi" => Service::ESXi,
            "snmp" => Service::SNMP,
            "krb5" => Service::KRB5,
            value => return Err(value.to_string()),
        })
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
/// Enum representing the type of credentials.
pub enum CredentialType {
    #[cfg_attr(feature = "serde_support", serde(rename = "up"))]
    /// User/password credentials.
    UP {
        /// The username for authentication.
        username: String,
        /// The password for authentication.
        password: String,
        /// privilege credential only use for SSH service
        #[cfg_attr(
            feature = "serde_support",
            serde(default, flatten, skip_serializing_if = "Option::is_none")
        )]
        privilege: Option<PrivilegeInformation>,
    },
    #[cfg_attr(feature = "serde_support", serde(rename = "usk"))]
    /// User/ssh-key credentials.
    USK {
        /// The username for authentication.
        username: String,
        /// The password for authentication.
        // A key without passphrase can be expected
        #[cfg_attr(
            feature = "serde_support",
            serde(default, skip_serializing_if = "Option::is_none")
        )]
        password: Option<String>,
        #[cfg_attr(feature = "serde_support", serde(rename = "private"))]
        /// The private key for authentication.
        private_key: String,
        /// privilege credential only use for SSH service
        #[cfg_attr(
            feature = "serde_support",
            serde(default, flatten, skip_serializing_if = "Option::is_none")
        )]
        privilege: Option<PrivilegeInformation>,
    },
    #[cfg_attr(feature = "serde_support", serde(rename = "snmp"))]
    /// SNMP credentials.
    SNMP {
        /// The SNMP username.
        username: String,
        /// The SNMP password.
        password: String,
        /// The SNMP community string.
        community: String,
        /// The SNMP authentication algorithm.
        auth_algorithm: String,
        /// The SNMP privacy password.
        privacy_password: String,
        /// The SNMP privacy algorithm.
        privacy_algorithm: String,
    },
    KRB5 {
        username: String,
        password: String,
        realm: String,
        kdc: String,
    },
}

impl CredentialType {
    /// Uses given closure to transform the password of the credential.
    pub fn map_password<F, E>(self, f: F) -> Result<Self, E>
    where
        F: Fn(String) -> Result<String, E>,
    {
        Ok(match self {
            CredentialType::UP {
                username,
                password,
                privilege,
            } => CredentialType::UP {
                username,
                password: f(password)?,
                privilege: match privilege {
                    Some(p) => Some(PrivilegeInformation {
                        username: p.username,
                        password: f(p.password)?,
                    }),
                    None => None,
                },
            },
            CredentialType::USK {
                username,
                password,
                private_key,
                privilege,
            } => CredentialType::USK {
                username,
                password: match password {
                    Some(p) => Some(f(p)?),
                    None => None,
                },
                private_key: f(private_key)?,
                privilege: match privilege {
                    Some(p) => Some(PrivilegeInformation {
                        username: p.username,
                        password: f(p.password)?,
                    }),
                    None => None,
                },
            },
            CredentialType::SNMP {
                username,
                password,
                community,
                auth_algorithm,
                privacy_password,
                privacy_algorithm,
            } => CredentialType::SNMP {
                username,
                password: f(password)?,
                community,
                auth_algorithm,
                privacy_password: f(privacy_password)?,
                privacy_algorithm,
            },
            CredentialType::KRB5 {
                username,
                password,
                realm,
                kdc,
            } => CredentialType::KRB5 {
                username,
                password: f(password)?,
                realm,
                kdc,
            },
        })
    }
}

impl AsRef<str> for CredentialType {
    fn as_ref(&self) -> &str {
        match self {
            CredentialType::UP { .. } => "up",
            CredentialType::USK { .. } => "usk",
            CredentialType::SNMP { .. } => "snmp",
            CredentialType::KRB5 { .. } => "krb5",
        }
    }
}
