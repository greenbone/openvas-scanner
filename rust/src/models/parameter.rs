// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

use std::fmt::Display;

#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
/// Represents a parameter for a VTS configuration.
pub struct Parameter {
    /// The ID of the parameter.
    pub id: u16,
    /// The value of the parameter.
    pub value: String,
}

impl Display for Parameter {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}: {}", self.id, self.value)
    }
}
