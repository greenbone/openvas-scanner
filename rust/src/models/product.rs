// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

/// Represents an product json file for notus
#[cfg_attr(feature = "serde_support", derive(serde::Deserialize))]
#[derive(Debug)]
pub struct Product {
    /// Version of the file, some version might not be supported by notus
    pub version: String,
    /// Package type, important for parsing the corresponding package
    pub package_type: PackageType,
    /// List of vulnerability tests for product
    #[cfg_attr(feature = "serde_support", serde(rename = "advisories"))]
    pub vulnerability_tests: Vec<VulnerabilityTest>,
}

/// Enum of supported package types
#[cfg_attr(feature = "serde_support", derive(serde::Deserialize))]
#[derive(Debug)]
pub enum PackageType {
    #[cfg_attr(feature = "serde_support", serde(rename = "deb"))]
    DEB,
    #[cfg_attr(feature = "serde_support", serde(rename = "ebuild"))]
    EBUILD,
    #[cfg_attr(feature = "serde_support", serde(rename = "rpm"))]
    RPM,
    #[cfg_attr(feature = "serde_support", serde(rename = "slack"))]
    SLACK,
    #[cfg_attr(feature = "serde_support", serde(rename = "msp"))]
    MSP,
    #[cfg_attr(feature = "serde_support", serde(rename = "alpm"))]
    ALPM,
}

/// Representing a single Vulnerability Test entry
#[cfg_attr(feature = "serde_support", derive(serde::Deserialize))]
#[derive(Debug)]
pub struct VulnerabilityTest {
    /// OID to identify vulnerability
    pub oid: String,
    /// List of affected packages, including the fixed version
    pub fixed_packages: Vec<FixedPackage>,
}

/// Version entry
#[cfg_attr(feature = "serde_support", derive(serde::Deserialize), serde(untagged))]
#[derive(Debug)]
pub enum FixedPackage {
    ByFullName {
        full_name: String,
        specifier: Specifier,
    },
    /// Contains a version and a specifier
    ByNameAndFullVersion {
        name: String,
        full_version: String,
        specifier: Specifier,
    },

    /// Contains a version Range
    ByRange { name: String, range: Range },
}

/// A specifier can be one of: >, <, >=, <=, =
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Deserialize, serde::Serialize)
)]
#[derive(Clone, Debug)]
pub enum Specifier {
    /// >
    #[cfg_attr(feature = "serde_support", serde(rename = ">"))]
    GT,
    /// <
    #[cfg_attr(feature = "serde_support", serde(rename = "<"))]
    LT,
    /// >=
    #[cfg_attr(feature = "serde_support", serde(rename = ">="))]
    GE,
    /// <=
    #[cfg_attr(feature = "serde_support", serde(rename = "<="))]
    LE,
    /// =
    #[cfg_attr(feature = "serde_support", serde(rename = "="))]
    EQ,
}

/// Version range
#[cfg_attr(feature = "serde_support", derive(serde::Deserialize))]
#[derive(Debug)]
pub struct Range {
    pub start: String,
    pub end: String,
}
