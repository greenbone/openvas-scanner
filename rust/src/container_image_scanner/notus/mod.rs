use std::{collections::HashMap, sync::Arc};

use greenbone_scanner_framework::models::{self, FixedVersion, VulnerablePackage};
use tokio::sync::RwLock;

use crate::{
    container_image_scanner::detection::OperatingSystem,
    notus::{HashsumProductLoader, Notus},
};

/// Some products have unique requirements that we have to generate somehow.
///
/// Some examples:
/// - NAME="openEuler" VERSION="24.03 (LTS-SP1)" -> openeuler_24.03_lts_sp1
fn generate_key(architecture: &str, os: &OperatingSystem) -> String {
    let normalize_euler_version = || {
        os.version
            .to_lowercase()
            .trim()
            .replace(" ", "_")
            .replace("(", "")
            .replace(")", "")
            .replace("-", "_")
    };
    let nos = os.name.to_lowercase();

    match (architecture, &nos as &str) {
        (_, "openeuler") => format!("openeuler_{}", normalize_euler_version()),
        (_, name) => format!("{}_{}", name, os.version_id),
    }
}

type NotusResults = HashMap<String, Vec<VulnerablePackage>>;

fn to_result(image: String, results: NotusResults) -> Vec<models::Result> {
    let hostname = Some(image);
    results
        .iter()
        .enumerate()
        .map(|(id, (oid, v))| {
            let messages = v
                .iter()
                .flat_map(|p| {
                    let versions = match &p.fixed_version {
                        FixedVersion::Single { version, specifier } => vec![format!(
                            "Fixed version:      {}{}-{}",
                            specifier, p.name, version
                        )],
                        FixedVersion::Range { start, end } => vec![
                            format!("Fixed version:      <={}-{}", p.name, start),
                            format!("Fixed version:      >={}-{}", p.name, end),
                        ],
                    };
                    let mut result = vec![
                        format!("Vulnerable package: {}", p.name),
                        format!("Installed version:  {}-{}", p.name, p.installed_version),
                    ];
                    result.extend(versions);
                    result
                })
                .collect::<Vec<_>>();
            models::Result {
                id,
                r_type: models::ResultType::Alarm,
                ip_address: None,
                hostname: hostname.clone(),
                oid: Some(oid.clone()),
                port: None,
                protocol: None,
                message: Some(messages.join("\n")),
                detail: None,
            }
        })
        .collect()
}

#[derive(Clone, Debug, PartialEq, Eq, thiserror::Error)]
pub enum Error {
    #[error("Unexpected response format from Notus.")]
    UnExpectedResponseFormat,
}

type Oz = Notus<HashsumProductLoader>;

pub async fn vulnerabilities(
    products: Arc<RwLock<Oz>>,
    architecture: &str,
    image: String,
    os: &OperatingSystem,
    packages: Vec<String>,
) -> Result<Vec<models::Result>, Error> {
    let mut p = products.write_owned().await;
    let os = generate_key(architecture, os);

    //TODO: here jo
    let result = tokio::task::spawn_blocking(move || p.scan(&os, &packages))
        .await
        .unwrap();
    match result {
        Ok(x) => Ok(to_result(image, x)),
        Err(error) => {
            tracing::warn!(%error, "Unable to get results from Notus.");
            Err(Error::UnExpectedResponseFormat)
        }
    }
    //
    // let request = Request::new(architecture, os, packages);
    // let response = request.send(config).await?;
    // Ok(to_result(image, response))
}

#[cfg(test)]
mod key_generation_tests {
    use crate::container_image_scanner::detection::OperatingSystem;

    #[test]
    fn debian() {
        let os = OperatingSystem {
            name: "debian".to_owned(),
            version: "12 (bookworm)".to_owned(),
            version_id: "12".to_owned(),
        };
        let result = super::generate_key("", &os);
        assert_eq!("debian_12".to_owned(), result);
    }

    #[test]
    fn openeuler() {
        let os = OperatingSystem {
            name: "openEuler".to_owned(),
            version: "24.03 (LTS-SP1)".to_owned(),
            version_id: "24.03".to_owned(),
        };
        let result = super::generate_key("", &os);
        assert_eq!("openeuler_24.03_lts_sp1".to_owned(), result);
    }
}

#[cfg(test)]
mod notus_result_parsing_tests {

    #[test]
    fn parse_range_and_fixed() {
        let json = r#"
        {
            "1.1": [
                {
                    "name": "foo",
                    "installed_version": "1.2.3",
                    "fixed_version": {
                        "start": "1.2.2",
                        "end": "1.2.5"
                    }
                },
                {
                    "name": "bar",
                    "installed_version": "1.2.4",
                    "fixed_version": {
                        "version": "1.2.5",
                        "specifier": ">="
                    }
                }
            ]
        }
        "#;
        let results: super::NotusResults = serde_json::from_str(json).unwrap();
        assert_eq!(results.len(), 1);
        let result = super::to_result("oci://holla/die:waldfee".to_owned(), results);
        insta::assert_ron_snapshot!(result);
    }
}
