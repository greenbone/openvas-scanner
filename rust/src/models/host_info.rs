// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

use std::collections::HashMap;

use crate::models::Host;

#[derive(Default)]
pub struct HostInfoBuilder {
    pub all: u64,
    pub excluded: u64,
    pub dead: u64,
    pub alive: u64,
    pub queued: u64,
    pub finished: u64,
    pub scanning: Option<HashMap<String, SingleHostScanInfo>>,
}

impl HostInfoBuilder {
    pub fn build(self) -> HostInfo {
        HostInfo {
            all: self.all,
            excluded: self.excluded,
            dead: self.dead,
            alive: self.alive,
            queued: self.queued,
            finished: self.finished,
            scanning: self.scanning,
            remaining_vts_per_host: HashMap::new(),
        }
    }
}

#[derive(Default, Debug, Clone, Eq, PartialEq)]
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
pub struct SingleHostScanInfo {
    finished_tests: AmountOfTests,
    total_tests: AmountOfTests,
}

#[derive(Debug, Clone, Eq, PartialEq)]
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
enum AmountOfTests {
    AmountOfTests(i32),
    DeadHost,
}

impl Default for AmountOfTests {
    fn default() -> Self {
        AmountOfTests::AmountOfTests(0)
    }
}

impl SingleHostScanInfo {
    pub fn new(finished_tests: i32, total_tests: i32) -> Self {
        Self {
            finished_tests: AmountOfTests::AmountOfTests(finished_tests),
            total_tests: if total_tests == -1 {
                AmountOfTests::DeadHost
            } else {
                AmountOfTests::AmountOfTests(total_tests)
            },
        }
    }

    pub fn is_finished(&self) -> bool {
        if let AmountOfTests::AmountOfTests(f) = self.finished_tests {
            if let AmountOfTests::AmountOfTests(t) = self.total_tests {
                return f == t;
            }
        }
        false
    }

    pub fn is_dead(&self) -> bool {
        matches!(self.total_tests, AmountOfTests::DeadHost)
    }
}

/// Information about hosts of a running scan
#[derive(Debug, Clone, Default, PartialEq, Eq)]
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
pub struct HostInfo {
    all: u64,
    excluded: u64,
    dead: u64,
    alive: u64,
    queued: u64,
    finished: u64,
    // Hosts that are currently being scanned. The second entry is the host
    // scan progress. Required for Openvas Scanner type
    #[cfg_attr(
        feature = "serde_support",
        serde(skip_serializing_if = "Option::is_none")
    )]
    scanning: Option<HashMap<String, SingleHostScanInfo>>,
    #[cfg_attr(feature = "serde_support", serde(skip))]
    remaining_vts_per_host: HashMap<String, usize>,
}

impl HostInfo {
    pub fn from_hosts_and_num_vts<'a>(
        targets: impl Iterator<Item = &'a str>,
        num_vts: usize,
    ) -> Self {
        let hosts: HashMap<_, _> = targets
            .map(|target| (target.to_string(), num_vts))
            .collect();
        Self {
            all: hosts.len() as u64,
            queued: hosts.len() as u64,
            remaining_vts_per_host: hosts,
            ..Default::default()
        }
    }

    pub fn register_finished_script(&mut self, target: &Host) {
        if let Some(num_vts) = self.remaining_vts_per_host.get_mut(target) {
            *num_vts -= 1;
            if *num_vts == 0 {
                self.finished += 1;
                self.queued -= 1;
                self.remaining_vts_per_host.remove(target);
            }
        }
    }

    pub fn finish(&mut self) {
        self.remaining_vts_per_host.clear();
        assert_eq!(self.queued, 0);
    }

    pub fn queued(&self) -> u64 {
        self.queued
    }

    pub fn finished(&self) -> u64 {
        self.finished
    }

    pub fn update_with(mut self, other: &HostInfo) -> Self {
        // total hosts value is sent once and only once must be updated
        if other.all != 0 {
            self.all = other.all;
        }
        // excluded hosts value is sent once and only once must be updated
        if self.excluded == 0 {
            self.excluded = other.excluded;
        }
        // new dead/alive/finished hosts are found during the scan.
        // the new count must be added to the previous one
        self.dead += other.dead;
        self.alive += other.alive;
        self.finished += other.finished;

        // Update each single host status. Remove it if finished.
        // Openvas doesn't keep the previous progress. Therefore
        // the values already stored in Openvasd must be updated
        // and never completely replaced.
        let mut hs = other.scanning.clone().unwrap_or_default();
        for (host, progress) in self.scanning.clone().unwrap_or_default().iter() {
            if progress.is_finished() || progress.is_dead() {
                hs.remove(host);
            } else {
                hs.insert(host.to_string(), progress.clone());
            }
        }
        self.scanning = Some(hs);
        self
    }
}
