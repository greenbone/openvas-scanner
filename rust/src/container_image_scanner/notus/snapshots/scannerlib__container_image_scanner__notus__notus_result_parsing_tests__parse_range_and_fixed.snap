---
source: src/container_image_scanner/notus/mod.rs
expression: result
---
[
  Result(
    id: 0,
    type: alarm,
    hostname: Some("oci://holla/die:waldfee"),
    oid: Some("1.1"),
    message: Some("Vulnerable package: foo\nInstalled version:  foo-1.2.3\nFixed version:      <=foo-1.2.2\nFixed version:      >=foo-1.2.5\nVulnerable package: bar\nInstalled version:  bar-1.2.4\nFixed version:      >=bar-1.2.5"),
  ),
]
